test/unittests/t_libcvmfs.cc

```diff
@@ -238,3 +251,217 @@ TEST_F(T_Libcvmfs, Attachv2) {
+                        "1dcca23355272056f04fe8bf20edfce0000000",
+                        "11111111111111111111111111111111111111"};
+
+const size_t file_size = 4096;
+
+/* Create directory specification for later repositories */
+DirSpec MakeBaseSpec() {
+  DirSpec spec;
+
+  // adding "/dir"
+  EXPECT_TRUE(spec.AddDirectory("dir", "", file_size));
+
+  // adding "/dir/file1"
+  EXPECT_TRUE(spec.AddFile("file1", "dir", hashes[0], file_size));
+
+  // adding "/dir/dir"
+  EXPECT_TRUE(spec.AddDirectory("dir",  "dir", file_size));
+  EXPECT_TRUE(spec.AddDirectory("dir2", "dir", file_size));
+  EXPECT_TRUE(spec.AddDirectory("dir3", "dir", file_size));
+
+  // adding "/dir/dir/file2"
+  EXPECT_TRUE(spec.AddFile("file2", "dir/dir", hashes[1], file_size));
+
+  // adding "/dir/dir2/file2"
+  EXPECT_TRUE(spec.AddFile("file2", "dir/dir2", hashes[3], file_size));
+
+  // adding "/dir/dir3/file2"
+  EXPECT_TRUE(spec.AddFile("file2", "dir/dir3", hashes[4], file_size));
+
+  // adding "/file3"
+  EXPECT_TRUE(spec.AddFile("file3", "", hashes[2], file_size));
+
+  return spec;
+}
+
+
+TEST_F(T_Libcvmfs, Listdir) {
+  /* Initialize options */
+  cvmfs_option_map *opts = cvmfs_options_init();
+
+  /* Create and initialize repository named "list" */
+  CatalogTestTool tester("list");
+  EXPECT_TRUE(tester.Init());
+
+  /* Create file structure */
+  DirSpec spec1 = MakeBaseSpec();
+  EXPECT_TRUE(tester.ApplyAtRootHash(tester.manifest()->catalog_hash(), spec1));
+
+  /* Set CVMFS options to reflect created repository */
+  cvmfs_options_set(opts, "CVMFS_ROOT_HASH",
+                    tester.manifest()->catalog_hash().ToString().c_str());
+  cvmfs_options_set(opts, "CVMFS_SERVER_URL",
+                    ("file://" + tester.repo_name()).c_str());
+  cvmfs_options_set(opts, "CVMFS_HTTP_PROXY", "DIRECT");
+  cvmfs_options_set(opts, "CVMFS_PUBLIC_KEY",
+                    tester.public_key().c_str());
+  cvmfs_options_set(opts, "CVMFS_CACHE_DIR",
+                    (tester.repo_name()+"/data/txn").c_str());
+  cvmfs_options_set(opts, "CVMFS_MOUNT_DIR",
+                          ("/cvmfs" + tester.repo_name()).c_str());
+
+  /* Initialize client repo based on options */
+  ASSERT_EQ(LIBCVMFS_ERR_OK, cvmfs_init_v2(opts));
+
+  /* Attach to client repo */
+  cvmfs_context *ctx;
+  EXPECT_EQ(LIBCVMFS_ERR_OK,
+      cvmfs_attach_repo_v2((tester.repo_name().c_str()), opts, &ctx));
+
+  char **buf = NULL;
+  size_t buflen = 0;
+  cvmfs_listdir(ctx, "/dir/dir", &buf, &buflen);
+  // The buf length should be at least 3
+  EXPECT_LE(3, buflen);
+  // Check that listed info matches specified dir
+  EXPECT_TRUE(!strcmp(".",  buf[0]));
+  EXPECT_TRUE(!strcmp("..", buf[1]));
+  EXPECT_TRUE(!strcmp("file2",  buf[2]));
+  // The 'last' value is null
+  EXPECT_TRUE(!buf[3]);
+
+  /* Finalize and close repo and options */
+  cvmfs_fini();
+  cvmfs_options_fini(opts);
+}
+
+TEST_F(T_Libcvmfs, StatNestedCatalog) {
```

Comment:
This should test finding deeply nested catalogs, too.  E.g. /nested/sub-nested in the case not even /nested has been attached yet.